### DISCUSSION TITLE
Why specify `--crush-location` when starting osd？

### DISCUSSION THREAD

**Discussion by microyahoo** [rook/rook#10769] [Q&A]
Users can create a custom `crushRoot` through the `ceph osd crush add-bucket` command, and then use this crushRoot in [pool spec](https://github.com/rook/rook/blob/73e0a864ce5942fe957e84f3442e2b902daac58d/pkg/apis/ceph.rook.io/v1/types.go#L618) to create an object store or ceph block pool.
> `crushRoot`: The root in the crush map to be used by the pool. If left empty or unspecified, the default root will be used. Creating a crush hierarchy for the OSDs currently requires the Rook toolbox to run the Ceph tools described [here](http://docs.ceph.com/docs/master/rados/operations/crush-map/#modifying-the-crush-map).

My questions are：
1. Since users can specify custom `crushRoot`, why specify `crush-location` when creating osd, is this redundant? I modified the code to remove this parameter, and no exceptions have been found so far. The following is original command.
```
ceph-osd --foreground --id 2 --fsid <fsid> --setuser ceph --setgroup ceph --crush-location=root=default host=<host>
```
https://github.com/rook/rook/blob/73e0a864ce5942fe957e84f3442e2b902daac58d/pkg/operator/ceph/cluster/osd/spec.go#L396

2. When the osd is started, this parameter `--crush-location` cannot be modified. If the user uses a custom `crushRoot`, this parameter  will be inconsistent with the parameter actually used by the user. Will there be any impact?

**Comment by travisn** [comment]
IIRC the crush location is only used by Ceph the first time the OSD starts. After that, the crush location param will be ignored by Ceph. So if you have changed the crush location of an OSD, I would expect it to remain that way. I believe this would work as expected, you will just have to modify the crush location after the osd starts the first time.

**Reply by microyahoo** [reply]
Also, I noticed that if I specify the `osd crush update on start = false` parameter in `ceph.conf` through `rook-config-override` configmap, the host crush is not created like below. Why is host crush not created under `root default`?
```
🍺 /root/go/src/deeproute.ai/smd ☞ git:(define_test_conf) ✗ceph osd tree      
ID  CLASS  WEIGHT  TYPE NAME                                         STATUS  REWEIGHT  PRI-AFF
-3              0  root root-43a48a86-73a6-45c8-b77c-ce72e83463bc                             
-5              0      host k1-c3b58398-2b4f-4ac6-8215-69149bdf24b7                           
 0    hdd       0          osd.0                                         up   1.00000  1.00000
 2    hdd       0          osd.2                                         up   1.00000  1.00000
-4              0      host k2-eee2c7f2-ce2c-4fe9-84e9-e54f457ee544                           
 1    hdd       0          osd.1                                         up   1.00000  1.00000
 5    hdd       0          osd.5                                         up   1.00000  1.00000
-6              0      host k3-cd578b6a-049a-45b8-9d84-e1fd93b39a16                           
 3    hdd       0          osd.3                                         up   1.00000  1.00000
 4    hdd       0          osd.4                                         up   1.00000  1.00000
-1              0  root default 
```
Subsequently, all pgs will be distributed on the same osd. why is that?
```
🍺 /root/go/src/deeproute.ai/smd ☞ git:(define_test_conf) ✗ceph osd df            
ID  CLASS  WEIGHT  REWEIGHT  SIZE     RAW USE  DATA     OMAP  META    AVAIL    %USE  VAR   PGS  STATUS
 0    hdd       0   1.00000   50 GiB   23 MiB  7.4 MiB   0 B  16 MiB   50 GiB  0.05  0.98    0      up
 2    hdd       0   1.00000   50 GiB   23 MiB  7.4 MiB   0 B  16 MiB   50 GiB  0.05  0.99    0      up
 1    hdd       0   1.00000   50 GiB   25 MiB  7.4 MiB   0 B  17 MiB   50 GiB  0.05  1.06  128      up
 5    hdd       0   1.00000   50 GiB   23 MiB  7.4 MiB   0 B  16 MiB   50 GiB  0.05  0.99    0      up
 3    hdd       0   1.00000   50 GiB   23 MiB  7.4 MiB   0 B  16 MiB   50 GiB  0.04  0.98    0      up
 4    hdd       0   1.00000   50 GiB   23 MiB  7.4 MiB   0 B  16 MiB   50 GiB  0.05  1.00    0      up
                      TOTAL  300 GiB  141 MiB   44 MiB   0 B  96 MiB  300 GiB  0.05                   
MIN/MAX VAR: 0.98/1.06  STDDEV: 0.00
```
@travisn

**Reply by travisn** [reply]
You moved the OSDs under the other root, correct? After your change the crush location, I don't believe ceph updates it even if the `--crush-location` is set

**Reply by microyahoo** [reply]
yes. It comes to normal  after I set weight for osd crush. Thanks for your explanation.
